File: Scripts/generate-name-policy-data.swift
#!/usr/bin/env swift
// Dev-only generator for the pinned Unicode data behind `NamePolicy` (SomnioCore).
//
// Run manually (NOT in CI) whenever the pinned Unicode version is bumped:
//
//     swift Scripts/generate-name-policy-data.swift [inputDir] [outputDir]
//
// With no arguments it downloads the six source files from unicode.org at the pinned
// version into a temp dir and writes the generated Swift into Sources/SomnioCore/Generated.
// Pass `inputDir` to read already-downloaded files from disk instead.
//
// Sources (all read at the SAME pinned version):
//   - confusables.txt        (security/<v>/)   TR39 confusable prototype map
//   - IdentifierStatus.txt   (security/<v>/)   Identifier_Status=Allowed set
//   - IdentifierType.txt     (security/<v>/)   read for provenance; Allowed already folds it in
//   - Scripts.txt            (<v>/ucd/)        primary Script per code point
//   - ScriptExtensions.txt   (<v>/ucd/)        Script_Extensions per code point
//   - PropertyValueAliases.txt (<v>/ucd/)      sc short<->long alias map (unifies the two spellings)
//
// IMPORTANT: bump `NamePolicy.skeletonAlgorithmVersion` whenever you bump `unicodeDataVersion`
// (or otherwise regenerate this data). The stored `name_skeleton_version` is gated on
// `skeletonAlgorithmVersion`, NOT on the data version — a data bump without a version bump
// is a generator bug, because the startup backfill would never recompute stale skeletons.

import Foundation

let unicodeVersion = "15.1.0"

let arguments = CommandLine.arguments
let explicitInputDir = arguments.count > 1 ? arguments[1] : nil
let outputDir = arguments.count > 2 ? arguments[2] : "Sources/SomnioCore/Generated"

// MARK: - Source loading

func loadSources() throws -> [String: String] {
    let securityNames = ["confusables.txt", "IdentifierStatus.txt", "IdentifierType.txt"]
    let ucdNames = ["Scripts.txt", "ScriptExtensions.txt", "PropertyValueAliases.txt"]
    var contents: [String: String] = [:]
    if let inputDir = explicitInputDir {
        for name in securityNames + ucdNames {
            let path = (inputDir as NSString).appendingPathComponent(name)
            contents[name] = try String(contentsOfFile: path, encoding: .utf8)
        }
    } else {
        let securityBase = "https://www.unicode.org/Public/security/\(unicodeVersion)/"
        let ucdBase = "https://www.unicode.org/Public/\(unicodeVersion)/ucd/"
        for name in securityNames {
            contents[name] = try String(contentsOf: URL(string: securityBase + name)!, encoding: .utf8)
        }
        for name in ucdNames {
            contents[name] = try String(contentsOf: URL(string: ucdBase + name)!, encoding: .utf8)
        }
    }
    return contents
}

// MARK: - Line parsing helpers

/// Strips a trailing `# ...` comment and trims whitespace. Returns nil for blank/comment lines.
func dataFields(_ line: String) -> [String]? {
    let withoutComment = line.split(separator: "#", maxSplits: 1, omittingEmptySubsequences: false)[0]
    let trimmed = withoutComment.trimmingCharacters(in: .whitespaces)
    guard !trimmed.isEmpty else { return nil }
    return trimmed.split(separator: ";", omittingEmptySubsequences: false).map {
        $0.trimmingCharacters(in: .whitespaces)
    }
}

/// Parses `XXXX` or `XXXX..YYYY` into an inclusive `(start, end)` code-point pair.
func parseRange(_ field: String) -> (UInt32, UInt32)? {
    let parts = field.components(separatedBy: "..")
    guard let start = UInt32(parts[0], radix: 16) else { return nil }
    if parts.count == 2, let end = UInt32(parts[1], radix: 16) { return (start, end) }
    return (start, start)
}

func hex(_ value: UInt32) -> String {
    String(value, radix: 16, uppercase: true)
}

// MARK: - Generation

let sources = try loadSources()

// --- Script short<->long alias map (sc property) ---
var shortToLong: [String: String] = [:]
var longNames: Set<String> = []
for line in sources["PropertyValueAliases.txt"]!.split(separator: "\n", omittingEmptySubsequences: false) {
    guard let fields = dataFields(String(line)), fields.count >= 3, fields[0] == "sc" else { continue }
    let short = fields[1]
    let long = fields[2]
    shortToLong[short] = long
    longNames.insert(long)
}

func canonicalScript(_ token: String) -> String {
    if longNames.contains(token) { return token }
    if let long = shortToLong[token] { return long }
    return token
}

/// --- Confusables (drop identity mappings) ---
var confusableEntries: [(UInt32, [UInt32])] = []
for line in sources["confusables.txt"]!.split(separator: "\n", omittingEmptySubsequences: false) {
    guard let fields = dataFields(String(line)), fields.count >= 2 else { continue }
    guard let source = UInt32(fields[0], radix: 16) else { continue }
    let targets = fields[1].split(separator: " ").compactMap { UInt32($0, radix: 16) }
    guard !targets.isEmpty else { continue }
    if targets == [source] { continue }
    confusableEntries.append((source, targets))
}

confusableEntries.sort { $0.0 < $1.0 }

/// --- Primary Script ranges ---
var scriptRanges: [(UInt32, UInt32, String)] = []
for line in sources["Scripts.txt"]!.split(separator: "\n", omittingEmptySubsequences: false) {
    guard let fields = dataFields(String(line)), fields.count >= 2, let range = parseRange(fields[0]) else { continue }
    scriptRanges.append((range.0, range.1, canonicalScript(fields[1])))
    longNames.insert(canonicalScript(fields[1]))
}

scriptRanges.sort { $0.0 < $1.0 }

/// --- Script_Extensions ranges ---
var scriptExtensionRanges: [(UInt32, UInt32, [String])] = []
for line in sources["ScriptExtensions.txt"]!.split(separator: "\n", omittingEmptySubsequences: false) {
    guard let fields = dataFields(String(line)), fields.count >= 2, let range = parseRange(fields[0]) else { continue }
    let scripts = fields[1].split(separator: " ").map { canonicalScript(String($0)) }
    for script in scripts {
        longNames.insert(script)
    }
    scriptExtensionRanges.append((range.0, range.1, scripts))
}

scriptExtensionRanges.sort { $0.0 < $1.0 }

/// --- Identifier_Status=Allowed ranges ---
var allowedRanges: [(UInt32, UInt32)] = []
for line in sources["IdentifierStatus.txt"]!.split(separator: "\n", omittingEmptySubsequences: false) {
    guard let fields = dataFields(String(line)), fields.count >= 2, fields[1] == "Allowed", let range = parseRange(fields[0]) else { continue }
    allowedRanges.append(range)
}

allowedRanges.sort { $0.0 < $1.0 }

// --- Stable script id assignment ---
let sortedScriptNames = longNames.sorted()
var scriptID: [String: Int] = [:]
for (index, name) in sortedScriptNames.enumerated() {
    scriptID[name] = index
}

// MARK: - Emit

let header = """
// Generated by Scripts/generate-name-policy-data.swift -- DO NOT EDIT BY HAND.
// Unicode data version \(unicodeVersion). Regenerate and bump
// NamePolicy.skeletonAlgorithmVersion together on any Unicode bump.

"""

let confusableTable = confusableEntries.map { source, targets in
    "\(hex(source))>\(targets.map(hex).joined(separator: " "))"
}.joined(separator: ";")

let confusablesFile = """
\(header)
enum ConfusablesData {
    /// `source>t1 t2 t3;...` -- TR39 confusable prototype map (identity mappings dropped).
    /// Source is a single scalar; the target is one or more scalars.
    static let mappingTable = "\(confusableTable)"
}
"""

let scriptNamesLine = sortedScriptNames.joined(separator: ";")
let scriptRangesLine = scriptRanges.map { "\(hex($0.0)) \(hex($0.1)) \(scriptID[$0.2]!)" }.joined(separator: ";")
let scriptExtensionsLine = scriptExtensionRanges.map { start, end, scripts in
    "\(hex(start)) \(hex(end)) \(scripts.map { String(scriptID[$0]!) }.joined(separator: ","))"
}.joined(separator: ";")

let scriptFile = """
\(header)
enum ScriptData {
    /// `;`-separated canonical (long) script names; array index is the script id used below.
    static let scriptNames = "\(scriptNamesLine)"

    /// `start end id;...` (hex code points, decimal id) -- primary Script per range.
    static let scriptRanges = "\(scriptRangesLine)"

    /// `start end id,id,...;...` -- Script_Extensions per range (overrides the primary Script).
    static let scriptExtensionRanges = "\(scriptExtensionsLine)"
}
"""

let allowedLine = allowedRanges.map { "\(hex($0.0)) \(hex($0.1))" }.joined(separator: ";")
let identifierFile = """
\(header)
enum IdentifierProfileData {
    /// `start end;...` (hex) -- the Identifier_Status=Allowed set. A code point is Allowed iff its
    /// Identifier_Type is a subset of {Recommended, Inclusion}; IdentifierStatus.txt is that
    /// precomputed result (IdentifierType.txt is read for provenance only).
    static let allowedRanges = "\(allowedLine)"
}
"""

let fileManager = FileManager.default
try fileManager.createDirectory(atPath: outputDir, withIntermediateDirectories: true)
func write(_ contents: String, to name: String) throws {
    let path = (outputDir as NSString).appendingPathComponent(name)
    try (contents + "\n").write(toFile: path, atomically: true, encoding: .utf8)
    print("wrote \(path) (\(contents.utf8.count) bytes)")
}

let versionFile = """
\(header)
enum NamePolicyDataVersion {
    /// The Unicode version this committed data was generated from. `NamePolicy.unicodeDataVersion`
    /// must equal this; a test asserts the two agree so a regeneration that forgets to update
    /// `NamePolicy` (and bump `skeletonAlgorithmVersion`) fails the build.
    static let unicode = "\(unicodeVersion)"
}
"""

try write(confusablesFile, to: "ConfusablesData.swift")
try write(scriptFile, to: "ScriptData.swift")
try write(identifierFile, to: "IdentifierProfileData.swift")
try write(versionFile, to: "NamePolicyDataVersion.swift")

print("confusable entries: \(confusableEntries.count)")
print("script ranges: \(scriptRanges.count), extension ranges: \(scriptExtensionRanges.count), scripts: \(sortedScriptNames.count)")
print("allowed ranges: \(allowedRanges.count)")

File: packages/data/scripts/generate-name-policy-data.mjs
#!/usr/bin/env node
// Dev-only generator for the pinned Unicode data behind the name policy.
//
// Run manually (NOT in CI) whenever the pinned Unicode version is bumped:
//
//     node packages/data/scripts/generate-name-policy-data.mjs [inputDir] [outputDir]
//
// With no arguments it downloads the seven source files from unicode.org at the pinned version
// into a temp dir and writes the JSON tables into packages/data/data/name-policy. Pass `inputDir`
// to read already-downloaded files from disk instead.
//
// Sources (all read at the SAME pinned version):
//   - confusables.txt          (security/<v>/)   TR39 confusable prototype map
//   - IdentifierStatus.txt     (security/<v>/)   Identifier_Status=Allowed set
//   - IdentifierType.txt       (security/<v>/)   downloaded for provenance, never parsed
//   - Scripts.txt              (<v>/ucd/)        primary Script per code point
//   - ScriptExtensions.txt     (<v>/ucd/)        Script_Extensions per code point
//   - PropertyValueAliases.txt (<v>/ucd/)        sc short<->long alias map (unifies the two spellings)
//   - UnicodeData.txt          (<v>/ucd/)        General_Category per code point (letters, marks, digits)
//
// A Unicode bump changes computed skeletons, so it is a migration that recomputes every stored
// `name_skeleton` row, and SKELETON_ALGORITHM_VERSION goes up with it.

import { mkdtemp, mkdir, readFile, writeFile } from 'node:fs/promises'
import { tmpdir } from 'node:os'
import { join, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'

const UNICODE_VERSION = '15.1.0'
const SECURITY_FILES = ['confusables.txt', 'IdentifierStatus.txt', 'IdentifierType.txt']
const UCD_FILES = ['Scripts.txt', 'ScriptExtensions.txt', 'PropertyValueAliases.txt', 'UnicodeData.txt']
/** The general categories the name shape admits: letters, marks, and decimal digits. */
const NAME_CATEGORIES = new Set(['Lu', 'Ll', 'Lt', 'Lm', 'Lo', 'Mn', 'Mc', 'Me', 'Nd'])

const [, , explicitInputDir, explicitOutputDir] = process.argv
const outputDir =
  explicitOutputDir ?? resolve(fileURLToPath(new URL('../data/name-policy/', import.meta.url)))

async function loadSources() {
  const contents = {}
  if (explicitInputDir !== undefined) {
    for (const name of [...SECURITY_FILES, ...UCD_FILES]) {
      contents[name] = await readFile(join(explicitInputDir, name), 'utf8')
    }
    return contents
  }
  const directory = await mkdtemp(join(tmpdir(), 'somnio-unicode-'))
  const fetchInto = async (base, name) => {
    const response = await fetch(`${base}${name}`)
    if (!response.ok) throw new Error(`${base}${name}: HTTP ${response.status}`)
    const text = await response.text()
    await writeFile(join(directory, name), text)
    contents[name] = text
  }
  for (const name of SECURITY_FILES)
    await fetchInto(`https://www.unicode.org/Public/security/${UNICODE_VERSION}/`, name)
  for (const name of UCD_FILES)
    await fetchInto(`https://www.unicode.org/Public/${UNICODE_VERSION}/ucd/`, name)
  return contents
}

/** Cuts the line at its first `#` and trims. Returns undefined for blank/comment lines. */
function dataFields(line) {
  const withoutComment = line.split('#', 1)[0].trim()
  if (withoutComment.length === 0) return undefined
  return withoutComment.split(';').map((field) => field.trim())
}

/** Parses `XXXX` or `XXXX..YYYY` into an inclusive `[start, end]` code-point pair. */
function parseRange(field) {
  const [start, end] = field.split('..')
  const from = parseInt(start, 16)
  if (Number.isNaN(from)) return undefined
  const to = end === undefined ? from : parseInt(end, 16)
  return Number.isNaN(to) ? undefined : [from, to]
}

const hex = (value) => value.toString(16).toUpperCase()
const lines = (text) => text.split('\n')

const sources = await loadSources()

// --- Script short<->long alias map (sc property) ---
const shortToLong = new Map()
const longNames = new Set()
for (const line of lines(sources['PropertyValueAliases.txt'])) {
  const fields = dataFields(line)
  if (fields === undefined || fields.length < 3 || fields[0] !== 'sc') continue
  shortToLong.set(fields[1], fields[2])
  longNames.add(fields[2])
}
const canonicalScript = (token) => (longNames.has(token) ? token : (shortToLong.get(token) ?? token))

// --- Confusables (drop identity mappings) ---
const confusableEntries = []
for (const line of lines(sources['confusables.txt'])) {
  const fields = dataFields(line)
  if (fields === undefined || fields.length < 2) continue
  const source = parseInt(fields[0], 16)
  if (Number.isNaN(source)) continue
  const targets = fields[1]
    .split(' ')
    .map((token) => parseInt(token, 16))
    .filter((value) => !Number.isNaN(value))
  if (targets.length === 0) continue
  if (targets.length === 1 && targets[0] === source) continue
  confusableEntries.push([source, targets])
}
confusableEntries.sort((a, b) => a[0] - b[0])

// --- Primary Script ranges ---
const scriptRanges = []
for (const line of lines(sources['Scripts.txt'])) {
  const fields = dataFields(line)
  if (fields === undefined || fields.length < 2) continue
  const range = parseRange(fields[0])
  if (range === undefined) continue
  const script = canonicalScript(fields[1])
  scriptRanges.push([range[0], range[1], script])
  longNames.add(script)
}
scriptRanges.sort((a, b) => a[0] - b[0])

// --- Script_Extensions ranges ---
const scriptExtensionRanges = []
for (const line of lines(sources['ScriptExtensions.txt'])) {
  const fields = dataFields(line)
  if (fields === undefined || fields.length < 2) continue
  const range = parseRange(fields[0])
  if (range === undefined) continue
  const scripts = fields[1].split(' ').map(canonicalScript)
  for (const script of scripts) longNames.add(script)
  scriptExtensionRanges.push([range[0], range[1], scripts])
}
scriptExtensionRanges.sort((a, b) => a[0] - b[0])

// --- Identifier_Status=Allowed ranges ---
const allowedRanges = []
for (const line of lines(sources['IdentifierStatus.txt'])) {
  const fields = dataFields(line)
  if (fields === undefined || fields.length < 2 || fields[1] !== 'Allowed') continue
  const range = parseRange(fields[0])
  if (range === undefined) continue
  allowedRanges.push(range)
}
allowedRanges.sort((a, b) => a[0] - b[0])

// --- General_Category ranges for the name-shape categories ---
// UnicodeData.txt lists one code point per line, except `<..., First>` / `<..., Last>` pairs that
// bracket a range. Adjacent code points with the same category are merged into one range.
const categoryRanges = []
let pendingFirst
let last
for (const line of lines(sources['UnicodeData.txt'])) {
  if (line.trim().length === 0) continue
  const fields = line.split(';')
  const code = parseInt(fields[0], 16)
  const name = fields[1]
  const category = fields[2]
  let start = code
  let end = code
  if (name.endsWith(', First>')) {
    pendingFirst = code
    continue
  }
  if (name.endsWith(', Last>')) {
    start = pendingFirst
    pendingFirst = undefined
  }
  if (!NAME_CATEGORIES.has(category)) continue
  if (last !== undefined && last.category === category && last.end + 1 === start) {
    last.end = end
    continue
  }
  last = { start, end, category }
  categoryRanges.push(last)
}

// --- Stable script id assignment ---
const sortedScriptNames = [...longNames].sort()
const scriptID = new Map(sortedScriptNames.map((name, index) => [name, index]))

const confusables = {
  mappingTable: confusableEntries
    .map(([source, targets]) => `${hex(source)}>${targets.map(hex).join(' ')}`)
    .join(';'),
}
const scripts = {
  scriptNames: sortedScriptNames.join(';'),
  scriptRanges: scriptRanges
    .map(([start, end, script]) => `${hex(start)} ${hex(end)} ${scriptID.get(script)}`)
    .join(';'),
  scriptExtensionRanges: scriptExtensionRanges
    .map(
      ([start, end, names]) =>
        `${hex(start)} ${hex(end)} ${names.map((name) => scriptID.get(name)).join(',')}`
    )
    .join(';'),
}
const identifierProfile = {
  allowedRanges: allowedRanges.map(([start, end]) => `${hex(start)} ${hex(end)}`).join(';'),
}
const generalCategories = {
  ranges: categoryRanges.map(({ start, end, category }) => `${hex(start)} ${hex(end)} ${category}`).join(';'),
}
const version = { unicode: UNICODE_VERSION }

await mkdir(outputDir, { recursive: true })
const write = async (name, value) => {
  const path = join(outputDir, name)
  await writeFile(path, `${JSON.stringify(value, null, 2)}\n`)
  console.log(`wrote ${path}`)
}
await write('confusables.json', confusables)
await write('scripts.json', scripts)
await write('identifier-profile.json', identifierProfile)
await write('general-categories.json', generalCategories)
await write('version.json', version)

console.log(`confusable entries: ${confusableEntries.length}`)
console.log(
  `script ranges: ${scriptRanges.length}, extension ranges: ${scriptExtensionRanges.length}, scripts: ${sortedScriptNames.length}`
)
console.log(`allowed ranges: ${allowedRanges.length}, category ranges: ${categoryRanges.length}`)
